File: matlab/gtfDeepCatalogger_ver1.m
function [Bcount,Ocount,Btable,Otable,Dtable] = gtfDeepCatalogger_ver1( ...
    v24, site, segN, fileN, rootdir, classifier, Frange, predffreq, predffreq_uncert, ...
    fstartdatetime, NSEC, varargin)

% Per-segment classifier & detector — fine-tuned network version.
% classifier must be a struct with fields:
%   .net     — fine-tuned DAGNetwork (ResNet50 head replaced for 2 classes)
%   .bestThr — decision threshold on P(bwhistle)
%   .posIdx  — column index of 'bwhistle' in the network's softmax output
%
% Options:
%   'SaveImages'     (true|false, default false)
%   'KeepImages'     ('all'|'bwhistle'|'other'|'none', default 'all')
%   'UseParInternal' (true|false, default false)

%% --- Parse options
p = inputParser;
p.addParameter('SaveImages', false, @(x)islogical(x)&&isscalar(x));
p.addParameter('KeepImages','all', @(s)ischar(s)||isstring(s));
p.addParameter('UseParInternal', false, @(x)islogical(x)&&isscalar(x));
p.parse(varargin{:});
SaveImages     = p.Results.SaveImages;
KeepImages     = lower(string(p.Results.KeepImages));
UseParInternal = p.Results.UseParInternal;

minute_start = (str2double(segN)-1)*(NSEC/60);

%% --- Detector params
s=10; sweep=6; ploton=0; thres=0.45;
fs=24000;
winLen   = round(1.1*fs);
half1Win = round(0.4*fs);
half2Win = winLen - half1Win;

%% --- Run matched-filter detector
[dettimesMF,ffreq,ccscore,l_detscore,u_detscore,fo_calls,f1_calls] = ...
    gtfMatchedFilterDet_ver1(v24,Frange,s,sweep,thres,predffreq,predffreq_uncert,ploton);

Btable = table(); Otable = table(); Dtable = table();
Bcount = 0; Ocount = 0;

if isempty(dettimesMF)
    [Btable,Otable,Dtable] = makeTablesEmpty();
    return
end

%% --- Timing conversions & F1 subset
DetInSamples = round(fs*dettimesMF);
DetInSamples(DetInSamples+half2Win >= numel(v24)) = max(1, numel(v24)-(winLen+1));
DetTime  = DetInSamples./fs;

if ~isdatetime(fstartdatetime)
    fstartdatetime = datetime(fstartdatetime,'TimeZone','UTC');
elseif isempty(fstartdatetime.TimeZone)
    fstartdatetime.TimeZone = 'UTC';
end
DetTime2 = fstartdatetime + seconds(DetTime);

isF1 = f1_calls & (l_detscore > thres * 1.1) & (ccscore > thres * 0.8);
DetInSamples_f1 = DetInSamples(isF1);
DetTime_f1      = DetTime(isF1);
DetTime2_f1     = DetTime2(isF1);
ffreq_f1        = ffreq(isF1);
ccscore_f1      = ccscore(isF1);
l_detscore_f1   = l_detscore(isF1);
u_detscore_f1   = u_detscore(isF1);

%% --- D table (all detections)
Dtable = table( ...
    repmat(string(site), numel(dettimesMF), 1), ...
    ffreq(:), ccscore(:), l_detscore(:), u_detscore(:), ...
    fo_calls(:), f1_calls(:), ...
    DetTime(:), DetTime2(:), ...
    repmat(string(segN), numel(dettimesMF), 1), ...
    repmat(minute_start, numel(dettimesMF), 1), ...
    repmat(string(fileN), numel(dettimesMF), 1), ...
    'VariableNames', {'Site','ffreq','ccscore','l_ccscore','u_ccscore', ...
                      'fo_calls','f1_calls','rel_time','abs_time','segN','seg_min','file'});
Dtable = sortrows(Dtable,"abs_time");
Dtable.abs_time.TimeZone = 'UTC';

if isempty(DetInSamples_f1)
    return
end

%% --- Fine-tuned network inference (batched)
persistent netFine
if isempty(netFine)
    netFine = classifier.net;
end
posIdx  = classifier.posIdx;
bestThr = classifier.bestThr;
tgtSize = [224 224]; batch = 128;

nF1     = numel(DetInSamples_f1);
kStarts = 1:batch:nF1;
nBlocks = numel(kStarts);
blocks  = cell(1, nBlocks);

for b = 1:nBlocks
    k1 = kStarts(b);
    k2 = min(k1+batch-1, nF1);
    blocks{b} = batchPredict( ...
        v24, DetInSamples_f1, k1, k2, half1Win, winLen, tgtSize, ...
        netFine, posIdx, UseParInternal);
end

pd_b = zeros(nF1, 1, 'single');
for b = 1:nBlocks
    k1 = kStarts(b);
    k2 = min(k1+batch-1, nF1);
    pd_b(k1:k2) = blocks{b};
end

%% --- Classify using threshold
labs   = categorical(double(pd_b) >= bestThr, [0 1], {'other','bwhistle'});
bb     = find(labs == 'bwhistle');
oo     = find(labs == 'other');
Bcount = numel(bb);
Ocount = numel(oo);

%% --- Build B/O tables
if Bcount > 0
    Btable = table( ...
        repmat(string(site), Bcount, 1), ...
        categorical(repmat("bwhistle", Bcount,1)), ...
        ffreq_f1(bb), ccscore_f1(bb), l_detscore_f1(bb), u_detscore_f1(bb), ...
        ones(Bcount,1), ...
        double(pd_b(bb)), ...
        DetTime_f1(bb), DetTime2_f1(bb), ...
        repmat(string(segN), Bcount, 1), ...
        repmat(minute_start, Bcount, 1), ...
        repmat(string(fileN), Bcount, 1), ...
        'VariableNames', {'Site','CallID','ffreq','ccscore','l_ccscore','u_ccscore', ...
                          'f1_calls','pdscore','rel_time','abs_time','segN','seg_min','file'});
    Btable = sortrows(Btable,"abs_time");
    Btable.abs_time.TimeZone = 'UTC';
else
    Btable = makeEmptyLike();
end

if Ocount > 0
    Otable = table( ...
        repmat(string(site), Ocount, 1), ...
        categorical(repmat("other", Ocount,1)), ...
        ffreq_f1(oo), ccscore_f1(oo), l_detscore_f1(oo), u_detscore_f1(oo), ...
        ones(Ocount,1), ...
        double(pd_b(oo)), ...
        DetTime_f1(oo), DetTime2_f1(oo), ...
        repmat(string(segN), Ocount, 1), ...
        repmat(minute_start, Ocount, 1), ...
        repmat(string(fileN), Ocount, 1), ...
        'VariableNames', {'Site','CallID','ffreq','ccscore','l_ccscore','u_ccscore', ...
                          'f1_calls','pdscore','rel_time','abs_time','segN','seg_min','file'});
    Otable = sortrows(Otable,"abs_time");
    Otable.abs_time.TimeZone = 'UTC';
else
    Otable = makeEmptyLike();
end

%% --- Optional: write JPGs per segment
if SaveImages
    keyStem        = regexprep(fileN, '\.[^.]+$', '');
    out_directory  = fullfile(rootdir, keyStem);
    out_directoryB = fullfile(out_directory, 'bwhistle');
    out_directoryO = fullfile(out_directory, 'other');
    if ~exist(out_directory,'dir'),  mkdir(out_directory);  end
    if ~exist(out_directoryO,'dir'), mkdir(out_directoryO); end
    if ~exist(out_directoryB,'dir'), mkdir(out_directoryB); end

    writeChosenImages(1:numel(DetInSamples_f1), out_directoryO, site, segN, fileN, minute_start, ...
        DetInSamples_f1, DetTime_f1, v24, half1Win, winLen, [224 224]);

    if any(KeepImages == ["all","bwhistle"]) && ~isempty(Btable)
        for k = 1:height(Btable)
            base = strcat(fileN, sprintf('%03.0f',minute_start), '_', sprintf('%011.7f',Btable.rel_time(k)));
            base = strrep(base, '.wav', '_');
            src  = fullfile(out_directoryO, [base, '.jpg']);
            dst  = fullfile(out_directoryB, [base, '.jpg']);
            if exist(src,'file'), movefile(src, dst, 'f'); end
        end
    end

    if KeepImages == "bwhistle"
        try, rmdir(out_directoryO,'s'); catch, end
    end
end

end

%%%% helpers %%%%

function pd_b = batchPredict(v24, DetInSamples_f1, k1, k2, half1Win, winLen, tgtSize, netFine, posIdx, useParImg)
nb        = k2 - k1 + 1;
imgs      = zeros([tgtSize 3 nb], 'uint8');
det_slice = DetInSamples_f1(k1:k2);
nv        = numel(v24);
if useParImg
    parfor j = 1:nb
        winStart = max(1, det_slice(j) - half1Win);
        winEnd   = min(nv, winStart + winLen - 1);
        if (winEnd - winStart + 1) < winLen, winStart = max(1, winEnd - winLen + 1); end
        im = gtfMakespectro_ver1(v24(winStart:winEnd), 0, 'OutFor', 'net');
        if size(im,3) == 1, im = repmat(im,1,1,3); end
        imgs(:,:,:,j) = im;
    end
else
    for j = 1:nb
        winStart = max(1, det_slice(j) - half1Win);
        winEnd   = min(nv, winStart + winLen - 1);
        if (winEnd - winStart + 1) < winLen, winStart = max(1, winEnd - winLen + 1); end
        im = gtfMakespectro_ver1(v24(winStart:winEnd), 0, 'OutFor', 'net');
        if size(im,3) == 1, im = repmat(im,1,1,3); end
        imgs(:,:,:,j) = im;
    end
end
scores = predict(netFine, imgs, 'MiniBatchSize', 64);
pd_b   = single(scores(:, posIdx));
end

function [Btab,Otab,Dtab] = makeTablesEmpty()
Btab = makeEmptyLike();
Otab = makeEmptyLike();
Dtab = table(string.empty, [], [], [], [], [], [], [], datetime.empty, string.empty, [], string.empty, ...
    'VariableNames', {'Site','ffreq','ccscore','l_ccscore','u_ccscore','fo_calls','f1_calls','rel_time','abs_time','segN','seg_min','file'});
Dtab.abs_time.TimeZone = 'UTC';
end

function T = makeEmptyLike()
varNames = {'Site','CallID','ffreq','ccscore','l_ccscore','u_ccscore', ...
            'f1_calls','pdscore','rel_time','abs_time','segN','seg_min','file'};
varTypes = {'string','categorical','double','double','double','double', ...
            'double','double','double','datetime','string','double','string'};
T = table('Size',[0 numel(varNames)], 'VariableTypes',varTypes, 'VariableNames',varNames);
T.CallID = categorical(T.CallID, {'bwhistle','other'});
T.abs_time.TimeZone = 'UTC';
end

function writeChosenImages(idx, outDir, site, segN, fileN, minute_start, ...
        DetInSamples_f1, DetTime_f1, v24, half1Win, winLen, tgtSize)

if isempty(idx), return; end
if ~exist(outDir,'dir'), mkdir(outDir); end

batchW = 512;
H = tgtSize(1); W = tgtSize(2);

for p1 = 1:batchW:numel(idx)
    p2 = min(p1+batchW-1, numel(idx));
    nb = p2 - p1 + 1;

    imgs  = zeros(H, W, 3, nb, 'uint8');
    names = strings(nb,1);
    [~, fileBase] = fileparts(fileN);

    for j = 1:nb
        g = idx(p1 + j - 1);
        winStart = max(1, DetInSamples_f1(g) - half1Win);
        winEnd   = min(numel(v24), winStart + winLen - 1);
        if (winEnd - winStart + 1) < winLen
            winStart = max(1, winEnd - winLen + 1);
        end
        ydet = v24(winStart:winEnd);

        im = gtfMakespectro_ver1(ydet, 0, ...
            'OutFor','file', ...
            'Scale','robust', ...
            'CMap', parula(256));
        imgs(:,:,:,j) = im;

        names(j) = fullfile(outDir, sprintf('%s_%03.0f_%011.7f.jpg', ...
                        fileBase, minute_start, DetTime_f1(g)));
    end

    for j = 1:nb
        imwrite(imgs(:,:,:,j), names(j), 'jpg', 'Quality', 90);
    end
end
end

File: python/gtfDeepCatalogger_ver1.py
"""
gtfDeepCatalogger_ver1.py  —  Python port of gtfDeepCatalogger_ver1.m

Per-segment matched-filter detector + deep-net (ResNet50) classifier.

classifier dict fields
----------------------
    net      : torch.nn.Module   (fine-tuned ResNet50, 2-class head)
    bestThr  : float             (decision threshold on P(bwhistle))
    posIdx   : int               (column index of bwhistle in softmax, = 1)

Returns
-------
Bcount, Ocount, Btable, Otable, Dtable
    Bcount / Ocount : int
    B/O/Dtable      : pandas.DataFrame (empty if nothing found)
"""

import io
import shutil
from pathlib import Path

import numpy as np
import pandas as pd
import torch
import torch.nn.functional as F
from PIL import Image
from torchvision import transforms

from gtfMakespectro_ver1 import gtf_make_spectro, gtf_load_parula_csv, _PARULA
from gtfMatchedFilterDet_ver1 import gtf_matched_filter_det

# ImageNet normalisation — must match training transforms in gtfDeepToadFinetuneScript.py
_MEAN = [0.485, 0.456, 0.406]
_STD  = [0.229, 0.224, 0.225]

_EVAL_TF = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(_MEAN, _STD),
])

# Load exact parula colormap once at import time.
# Must match what gtfParseTrainJpg.py used when building train_jpg_py — if
# parula256.csv was present then (it was), inference must use it too.
_PARULA_CSV = Path(__file__).parent / 'parula256.csv'
_CMAP = gtf_load_parula_csv(str(_PARULA_CSV)) if _PARULA_CSV.exists() else _PARULA
if not _PARULA_CSV.exists():
    import warnings
    warnings.warn('parula256.csv not found — using approximate colormap. '
                  'Classifier accuracy may be degraded.', RuntimeWarning)

# ---- column schemas -------------------------------------------------------
_BTAB_COLS = ['Site','CallID','ffreq','ccscore','l_ccscore','u_ccscore',
              'f1_calls','pdscore','rel_time','abs_time','segN','seg_min','file']
_DTAB_COLS = ['Site','ffreq','ccscore','l_ccscore','u_ccscore',
              'fo_calls','f1_calls','rel_time','abs_time','segN','seg_min','file']


def _empty_btab():
    return pd.DataFrame(columns=_BTAB_COLS)


def _empty_dtab():
    return pd.DataFrame(columns=_DTAB_COLS)


# ---- internal helpers -----------------------------------------------------

def _spectro_tensor(ydet: np.ndarray) -> torch.Tensor:
    """
    Spectrogram → normalised float32 tensor shape (3, 224, 224).

    Uses the exact parula256.csv colormap and a quality-95 JPEG round-trip
    so pixel values match what the network saw during training.
    """
    im, _ = gtf_make_spectro(ydet, out_for='net', cmap=_CMAP)  # uint8 (224, 224, 3)
    buf = io.BytesIO()
    Image.fromarray(im).save(buf, format='JPEG', quality=95)
    buf.seek(0)
    pil = Image.open(buf).convert('RGB')
    return _EVAL_TF(pil)                                        # (3, 224, 224)


def _batch_predict(v24, det_in_samples_f1, k1, k2,
                   half1Win, winLen, net, pos_idx, device):
    """
    Build spectrogram batch for samples[k1:k2], run net, return P(bwhistle).
    """
    det_slice = det_in_samples_f1[k1:k2]
    nv        = len(v24)

    tensors = []
    for samp in det_slice:
        win_start = max(0, samp - half1Win)
        win_end   = min(nv, win_start + winLen)
        if (win_end - win_start) < winLen:
            win_start = max(0, win_end - winLen)
        tensors.append(_spectro_tensor(v24[win_start:win_end]))

    batch = torch.stack(tensors).to(device)   # (nb, 3, 224, 224)
    net.eval()
    with torch.no_grad():
        logits = net(batch)
        probs  = F.softmax(logits, dim=1)[:, pos_idx].cpu().numpy()

    return probs.astype(np.float32)


def _write_images(v24, det_in_samples_f1, det_time_f1, half1Win, winLen,
                  out_dir_o, file_base, minute_start, parula_cmap):
    """Write all F1-candidate spectrograms as JPEGs into out_dir_o."""
    nv = len(v24)
    for j, samp in enumerate(det_in_samples_f1):
        win_start = max(0, samp - half1Win)
        win_end   = min(nv, win_start + winLen)
        if (win_end - win_start) < winLen:
            win_start = max(0, win_end - winLen)
        ydet = v24[win_start:win_end]

        im, _    = gtf_make_spectro(ydet, out_for='file', scale='robust', cmap=parula_cmap)
        img_name = f'{file_base}_{minute_start:03.0f}_{det_time_f1[j]:011.7f}.jpg'
        Image.fromarray(im).save(str(out_dir_o / img_name), format='JPEG', quality=90)


# ---- main function --------------------------------------------------------

def gtf_deep_catalogger(
    v24, site, segN, fileN, rootdir, classifier,
    Frange, predffreq, predffreq_uncert,
    fstartdatetime, NSEC,
    save_images: bool = False,
    keep_images: str  = 'all',
    device=None,
):
    """
    Run the matched-filter detector on one 2-min audio segment, then pass
    high-confidence f1 candidates through the fine-tuned ResNet50 classifier.

    Parameters
    ----------
    v24 : np.ndarray
        1-D audio at 24 kHz.
    site, segN, fileN : str
        Site name, zero-padded segment number ('001'), source filename.
    rootdir : str or Path
        Root output directory for JPEG images.
    classifier : dict
        {'net': nn.Module, 'bestThr': float, 'posIdx': int}
    Frange : (float, float)
        Frequency search range [f_low, f_high] Hz.
    predffreq : float
        Predicted fundamental frequency (Hz).
    predffreq_uncert : float
        Uncertainty (Hz).
    fstartdatetime : datetime or pd.Timestamp
        UTC start time of this segment.
    NSEC : float
        Segment duration (s).
    save_images : bool
    keep_images : str   'all' | 'bwhistle' | 'other' | 'none'
    device : torch.device or None

    Returns
    -------
    Bcount, Ocount, Btable, Otable, Dtable
    """
    if device is None:
        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    keep_images  = keep_images.lower()
    minute_start = (int(segN) - 1) * (NSEC / 60)

    # ---- detector params (must match MATLAB) ----------------------------
    s = 10; sweep = 6; ploton = False; thres = 0.45
    fs       = 24_000
    winLen   = round(1.1 * fs)    # 26400 samples
    half1Win = round(0.4 * fs)    # 9600  samples (lookback)

    v24 = np.asarray(v24, dtype=np.float64).ravel()

    # ---- matched-filter detector ----------------------------------------
    det_times_mf, ffreq, ccscore, l_detscore, u_detscore, fo_calls, f1_calls = \
        gtf_matched_filter_det(v24, Frange, s, sweep, thres,
                               predffreq, predffreq_uncert, ploton)

    if len(det_times_mf) == 0:
        return 0, 0, _empty_btab(), _empty_btab(), _empty_dtab()

    # ---- timing conversions ---------------------------------------------
    half2Win       = winLen - half1Win
    nv             = len(v24)
    det_in_samples = np.round(fs * det_times_mf).astype(int)

    # Clip detections too close to the end of the segment
    too_late = (det_in_samples + half2Win) >= nv
    det_in_samples[too_late] = max(1, nv - (winLen + 1))
    det_time = det_in_samples / fs

    if not isinstance(fstartdatetime, pd.Timestamp):
        fstartdatetime = pd.Timestamp(fstartdatetime, tz='UTC')
    elif fstartdatetime.tzinfo is None:
        fstartdatetime = fstartdatetime.tz_localize('UTC')
    det_time2 = [fstartdatetime + pd.Timedelta(seconds=float(t)) for t in det_time]

    # ---- D table (all detector candidates) ------------------------------
    n_det  = len(det_times_mf)
    Dtable = pd.DataFrame({
        'Site'      : [str(site)]   * n_det,
        'ffreq'     : ffreq,
        'ccscore'   : ccscore,
        'l_ccscore' : l_detscore,
        'u_ccscore' : u_detscore,
        'fo_calls'  : fo_calls.astype(bool),
        'f1_calls'  : f1_calls.astype(bool),
        'rel_time'  : det_time,
        'abs_time'  : det_time2,
        'segN'      : [str(segN)]   * n_det,
        'seg_min'   : [minute_start] * n_det,
        'file'      : [str(fileN)]  * n_det,
    })
    Dtable = Dtable.sort_values('abs_time').reset_index(drop=True)

    # ---- F1 subset: strong enough for deep-net scoring ------------------
    isF1              = f1_calls.astype(bool) & (l_detscore > thres * 1.1) & (ccscore > thres * 0.8)
    det_in_samples_f1 = det_in_samples[isF1]
    det_time_f1       = det_time[isF1]
    det_time2_f1      = [t for t, m in zip(det_time2, isF1) if m]
    ffreq_f1          = ffreq[isF1]
    ccscore_f1        = ccscore[isF1]
    l_detscore_f1     = l_detscore[isF1]
    u_detscore_f1     = u_detscore[isF1]

    if len(det_in_samples_f1) == 0:
        return 0, 0, _empty_btab(), _empty_btab(), Dtable

    # ---- batched deep-net inference -------------------------------------
    net      = classifier['net']
    pos_idx  = classifier['pos_idx']
    best_thr = classifier['bestThr']
    BATCH    = 128

    nF1  = len(det_in_samples_f1)
    pd_b = np.zeros(nF1, dtype=np.float32)
    for k1 in range(0, nF1, BATCH):
        k2 = min(k1 + BATCH, nF1)
        pd_b[k1:k2] = _batch_predict(
            v24, det_in_samples_f1, k1, k2, half1Win, winLen,
            net, pos_idx, device)

    # ---- classify -------------------------------------------------------
    is_bwhistle = pd_b >= best_thr
    bb = np.where(is_bwhistle)[0]
    oo = np.where(~is_bwhistle)[0]
    Bcount = len(bb)
    Ocount = len(oo)

    def _make_tab(idx_arr, label):
        n = len(idx_arr)
        if n == 0:
            return _empty_btab()
        df = pd.DataFrame({
            'Site'      : [str(site)]    * n,
            'CallID'    : [label]        * n,
            'ffreq'     : ffreq_f1[idx_arr],
            'ccscore'   : ccscore_f1[idx_arr],
            'l_ccscore' : l_detscore_f1[idx_arr],
            'u_ccscore' : u_detscore_f1[idx_arr],
            'f1_calls'  : np.ones(n, dtype=int),
            'pdscore'   : pd_b[idx_arr].astype(float),
            'rel_time'  : det_time_f1[idx_arr],
            'abs_time'  : [det_time2_f1[i] for i in idx_arr],
            'segN'      : [str(segN)]    * n,
            'seg_min'   : [minute_start] * n,
            'file'      : [str(fileN)]   * n,
        })
        return df.sort_values('abs_time').reset_index(drop=True)

    Btable = _make_tab(bb, 'bwhistle')
    Otable = _make_tab(oo, 'other')

    # ---- optional JPEG output -------------------------------------------
    if save_images:
        parula_cmap = _CMAP   # loaded at module import from parula256.csv

        file_base = Path(fileN).stem
        out_dir   = Path(rootdir) / file_base
        out_dir_o = out_dir / 'other'
        out_dir_b = out_dir / 'bwhistle'
        out_dir_o.mkdir(parents=True, exist_ok=True)
        out_dir_b.mkdir(parents=True, exist_ok=True)

        # Write all F1 candidates to 'other' folder first
        _write_images(v24, det_in_samples_f1, det_time_f1,
                      half1Win, winLen, out_dir_o, file_base,
                      minute_start, parula_cmap)

        # Move bwhistle classifications
        if keep_images in ('all', 'bwhistle') and Bcount > 0:
            for _, row in Btable.iterrows():
                img_name = f'{file_base}_{minute_start:03.0f}_{row["rel_time"]:011.7f}.jpg'
                src = out_dir_o / img_name
                dst = out_dir_b / img_name
                if src.exists():
                    src.rename(dst)

        if keep_images == 'bwhistle':
            try:
                shutil.rmtree(str(out_dir_o))
            except Exception:
                pass

    return Bcount, Ocount, Btable, Otable, Dtable
